include/alpaka/math/Complex.hpp

```diff
@@ -0,0 +1,540 @@
+    //! Additionally, provides an implicit conversion to and from std::complex<T>.
+    //! All methods besides operators << and >> are host-device.
+    //! Note that it does not provide non-member functions of std::complex besides the operators.
+    //! Those are provided the same way as alpaka math functions for real numbers.
+    //!
+    //! Note that unlike most of alpaka, this is a concrete type template, and not merely a concept.
+    //!
+    //! Naming and order of the methods match https://en.cppreference.com/w/cpp/numeric/complex.
+    //!
+    //! @tparam T type of the real and imaginary part: float, double, or long double.
+    template<typename T>
+    class Complex
+    {
+    public:
+        //! Type of the real and imaginary parts
+        using value_type = T;
+
+        //! Constructor from the given real and imaginary parts
+        constexpr ALPAKA_FN_HOST_ACC Complex(T const& real = T{}, T const& imag = T{}) : m_real(real), m_imag(imag)
+        {
+        }
+
+        //! Copy constructor
+        constexpr ALPAKA_FN_HOST_ACC Complex(Complex const& other) = default;
+
+        //! Constructor from Complex of another type
+        template<typename U>
+        constexpr ALPAKA_FN_HOST_ACC Complex(Complex<U> const& other)
+            : m_real(static_cast<T>(other.real()))
+            , m_imag(static_cast<T>(other.imag()))
+        {
+        }
+
+        //! Constructor from std::complex
+        constexpr ALPAKA_FN_HOST_ACC Complex(std::complex<T> const& other) : m_real(other.real()), m_imag(other.imag())
+        {
+        }
+
+        //! Conversion to std::complex
+        constexpr ALPAKA_FN_HOST_ACC operator std::complex<T>() const
+        {
+            return std::complex<T>{m_real, m_imag};
+        }
+
+        //! Assignment
+        ALPAKA_FN_HOST_ACC Complex& operator=(Complex const&) = default;
+
+        //! Get the real part
+        constexpr ALPAKA_FN_HOST_ACC T real() const
+        {
+            return m_real;
+        }
+
+        //! Set the real part
+        constexpr ALPAKA_FN_HOST_ACC void real(T value)
+        {
+            m_real = value;
+        }
+
+        //! Get the imaginary part
+        constexpr ALPAKA_FN_HOST_ACC T imag() const
+        {
+            return m_imag;
+        }
+
+        //! Set the imaginary part
+        constexpr ALPAKA_FN_HOST_ACC void imag(T value)
+        {
+            m_imag = value;
+        }
+
+        //! Addition assignment with a real number
+        ALPAKA_FN_HOST_ACC Complex& operator+=(T const& other)
+        {
+            m_real += other;
+            return *this;
+        }
+
+        //! Addition assignment with a complex number
+        template<typename U>
+        ALPAKA_FN_HOST_ACC Complex& operator+=(Complex<U> const& other)
+        {
+            m_real += static_cast<T>(other.real());
+            m_imag += static_cast<T>(other.imag());
+            return *this;
+        }
```

Comment:
```suggestion
        ALPAKA_FN_HOST_ACC Complex& operator+=(Complex const& other)
        {
            m_real += other.real();
            m_imag += other.imag();
            return *this;
        }
```
For `operator+=` you support converting from complex numbers with different FP type, but e.g. `operator+` does not support his conversion. I would keep it stricter and only allow a matching complex number in arithmetic operations.

Reply:
I re-checked, that's what `std::complex` does, only assignment-operation has mixed-type support

Reply:
Ok then!